components/script/dom/notification.rs

```diff
@@ -560,3 +636,331 @@ fn request_notification_permission(global: &GlobalScope, can_gc: CanGc) -> Notif
+                    resource_type.clone(),
+                );
+            },
+            ImageCacheResult::Pending(pending_image_id) => {
+                self.register_image_cache_callback(
+                    request_id,
+                    pending_image_id,
+                    resource_type.clone(),
+                );
+            },
+            ImageCacheResult::ReadyForRequest(pending_image_id) => {
+                self.register_image_cache_callback(
+                    request_id,
+                    pending_image_id,
+                    resource_type.clone(),
+                );
+                self.fetch(pending_image_id, request, global);
+            },
+            ImageCacheResult::LoadError => {
+                self.set_resource_and_show_when_ready(request_id, &resource_type, None);
+            },
+        };
+    }
+
+    fn register_image_cache_callback(
+        &self,
+        request_id: RequestId,
+        pending_image_id: PendingImageId,
+        resource_type: ResourceType,
+    ) {
+        let (sender, receiver) =
+            ipc::channel::<PendingImageResponse>().expect("ipc channel failure");
+
+        let global: &GlobalScope = &self.global();
+
+        let trusted_this = Trusted::new(self);
+        let resource_type = resource_type.clone();
+        let task_source = global.task_manager().networking_task_source().to_sendable();
+
+        ROUTER.add_typed_route(
+            receiver,
+            Box::new(move |response| {
+                let trusted_this = trusted_this.clone();
+                let resource_type = resource_type.clone();
+                task_source.queue(task!(handle_response: move || {
+                    let this = trusted_this.root();
+                    if let Ok(response) = response {
+                        this.handle_image_cache_response(request_id, response.response, resource_type);
+                    } else {
+                        this.handle_image_cache_response(request_id, ImageResponse::None, resource_type);
+                    }
+                }));
+            }),
+        );
+
+        global.image_cache().add_listener(ImageResponder::new(
+            sender,
+            global.pipeline_id(),
+            pending_image_id,
+        ));
+    }
+
+    fn handle_image_cache_response(
+        &self,
+        request_id: RequestId,
+        response: ImageResponse,
+        resource_type: ResourceType,
+    ) {
+        match response {
+            ImageResponse::Loaded(image, _) => {
+                self.set_resource_and_show_when_ready(request_id, &resource_type, Some(image));
+            },
+            ImageResponse::PlaceholderLoaded(image, _) => {
+                self.set_resource_and_show_when_ready(request_id, &resource_type, Some(image));
+            },
+            ImageResponse::None => {
+                self.set_resource_and_show_when_ready(request_id, &resource_type, None);
+            },
+            _ => (),
```

Comment:
Do other variants also need to be checked here?

Reply:
The last one is `MetadataLoaded`. I think we don't need it at the moment.